test/e2e/topology_updater.go

```diff
@@ -41,35 +44,154 @@ import (
+
 			topologyClient, err = topologyclientset.NewForConfig(f.ClientConfig())
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
-			topologyUpdaterNode, err = f.ClientSet.CoreV1().Nodes().Get(context.TODO(), getNodeName(), metav1.GetOptions{})
+			topologyUpdaterNode, err = f.ClientSet.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+			workerNodes, err = utils.GetWorkerNodes(f)
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
 			initialized = true
 		}
 
-		// intentionally get every single time
+		// intentionally done once
 		kubeletConfig, err = e2ekubelet.GetCurrentKubeletConfig(topologyUpdaterNode.Name, "", true)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 	})
 
 	ginkgo.Context("with cluster configured", func() {
+		ginkgo.It("it should not account for any cpus if a container doesn't request exclusive cpus", func() {
+			ginkgo.By("getting the initial topology information")
+			initialNodeTopo := getNodeTopology(topologyClient, topologyUpdaterNode.Name, namespace)
+			ginkgo.By("creating a pod consuming the shared pool")
+			sleeperPod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "sleeper-be-pod",
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						v1.Container{
+							Name:  "sleeper-be-cnt",
+							Image: utils.CentosImage,
+							// 1 hour (or >= 1h in general) is "forever" for our purposes
+							Command: []string{"/bin/sleep", "1h"},
+						},
+					},
+				},
+			}
+
+			podMap := make(map[string]*v1.Pod)
+			pod := f.PodClient().CreateSync(sleeperPod)
+			podMap[pod.Name] = pod
+			defer utils.DeletePodsAsync(f, podMap)
+
+			cooldown := 30 * time.Second
+			ginkgo.By(fmt.Sprintf("getting the updated topology - sleeping for %v", cooldown))
+			// the object, hance the resource version must NOT change, so we can only sleep
+			time.Sleep(cooldown)
+			ginkgo.By("checking the changes in the updated topology - expecting none")
+			finalNodeTopo := getNodeTopology(topologyClient, topologyUpdaterNode.Name, namespace)
+
+			initialAllocRes := allocatableResourceListFromNodeResourceTopology(initialNodeTopo)
+			finalAllocRes := allocatableResourceListFromNodeResourceTopology(finalNodeTopo)
+			if len(initialAllocRes) == 0 || len(finalAllocRes) == 0 {
+				ginkgo.Fail(fmt.Sprintf("failed to find allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
+			}
+			zoneName, resName, cmp, ok := cmpAllocatableResources(initialAllocRes, finalAllocRes)
+			framework.Logf("zone=%q resource=%q cmp=%v ok=%v", zoneName, resName, cmp, ok)
+			if !ok {
+				ginkgo.Fail(fmt.Sprintf("failed to compare allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
+			}
+			// this is actually a workaround. The proper solution is to wait for ALL the container requesting exclusive resources to be gone
```

Comment:
My understanding was that at a given time for a particular test the pod above is running, is that not true? Based on that the `initialAllocRes` and `finalAllocRes` should be the same as the pod acquires CPUs from the shared pool. I am a little confused by this comment here that there might be other exclusive CPUs running?

Reply:
The comment is misplaced. Depending on the (random, by design) order on which ginkgo runs the tests, a test which exclusively allocates CPUs may run before. I didn't find yet a clean way to wait for this pod to be completely gone (hence, releasing the exclusively allocated CPUs) before to end the test, so *this* test can run with some leftovers hanging around, which makes the accounting harder.

I will improve and/or move the comment to capture this information.

Reply:
Thanks!